contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,671 @@
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Spec is the specification of the behavior of the autoscaler.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.
+	// +optional
+	Spec VerticalPodAutoscalerSpec
+
+	// Status is the current information about the autoscaler.
+	// +optional
+	Status VerticalPodAutoscalerStatus
+}
+
+
+// VerticalPodAutoscalerSpec is the specification of the behavior of the autoscaler.
+type VerticalPodAutoscalerSpec {
+	// PodSelector is a label query that determines the set of pods
+	// controlled by the Autoscaler.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	// +optional
+	PodSelector string
+
+	// UpdatePolicy describes the rules on how changes are applied to the pods.
+	// +optional
+	UpdatePolicy PodUpdatePolicy
+
+	// ResourcePolicy controls how the autoscaler computes recommended resources.
+	// +optional
+	ResourcePolicy PodResourcePolicy
+}
+
+// VerticalPodAutoscalerStatus
+type VerticalPodAutoscalerStatus {
+	// Recommendation is the most recently computed amount of resources
+	// recommended by the autoscaler for the controlled pods.
+	// +optional
+	Recommendation RecommendedPodResources
+}
+
+// UpdateMode controls when autoscaler applies changes to the pod resoures.
+type UpdateMode string
+const (
+	// UpdateModeOff means that autoscaler never changes Pod resources.
+	// The recommender still sets the recommended resources in the
+	// VerticalPodAutoscaler object. This can be used for a "dry run".
+	UpdateModeOff UpdateMode = "Off"
+	// UpdateModeInitial means that autoscaler only assigns resources on pod
+	// creation and does not change them during the lifetime of the pod.
+	UpdateModeInitial UpdateMode = "Initial"
+	// UpdateModeAuto means that autoscaler assigns resources on pod creation
+	// and additionally can update them during the lifetime of the pod,
+	// including evicting / rescheduling the pod.
+	UpdateModeAuto UpdateMode = "Auto"
+)
+
+// PodUpdatePolicy describes the rules on how changes are applied to the pods.
+type PodUpdatePolicy struct {
+	// UpdateMode controls when autoscaler applies changes to the pod resoures.
+	// +optional
+	UpdateMode UpdateMode
+}
+
+// ContainerResourcePolicy controls how autoscaler computes the recommended
+// resources for a specific container.
+type ContainerResourcePolicy struct {
+	// MinAllowed specifies the minimal amount of resources that will be
+	// recommended for the container.
+	// +optional
+	MinAllowed api.ResourceRequirements
+	// MaxAllowed specifies the maximum amount of resources that will be
+	// recommended for the container.
+	// +optional
+	MaxAllowed api.ResourceRequirements
+}
+
+// PodResourcePolicy controls how autoscaler computes the recommended resources
+// for containers belonging to the pod.
+type PodResourcePolicy struct {
```

Comment:
Strict mode still not added.

Reply:
I was thinking about it and frankly I'm a bit reluctant to add this knob before any user even requests it. We would have multiple settings controlling whether VPA is active: UpdateMode + per-container ResourcePolicy + "strict mode". Could we postpone it until we get more clarity whether it is really needed and in what form? We can also discuss more offline.

Reply:
I agree with @kgrygiel on this point, KISS rules. Let's figure out what users need first. I suppose it's OK to mention potential policies in the design doc and maybe encourage users to provide feedback based on their workloads/usage but yeah, let's remove this for now.

Reply:
What I ask for is not an extra shiny feature but a way for user to express what to do with containers for which min/max memory/cpu is not provided. 

Should it be unbounded or not autoscaled at all? Currently there is no convenient way to tell VPA to scale just one container in a Pod. For example scale my web server but leave metrics pusher alone, or any extra container the user may add in the future.

Reply:
OK, that sounds sensible @mwielgus and thanks for clarifying it! Why not, rather than making it configurable, choose a default behaviour for now (maybe non-auto-scaling, but no strong opinion here) and gather feedback from users as discussed above?

Reply:
Yes, if we go for no-autoscaling by default then we can skip the field for now.

Reply:
I'm reluctant to making it no-autoscaling by default, I would strive for the minimal working configuration being really minimal.
After further discussion we agreed to add an (optional) default ContainerResourcePolicy, which addresses the above concerns, while not complicating the API too much, I think.